### PR TITLE
[MLIR][ROCDL] Remove patterns for ops supported as intrinsics in the AMDGPU backend

### PR DESCRIPTION
**Pull request by jsjodin**
This patch removes patterns for a few operations which allows mathToLLVM conversion to convert the operations into LLVM intrinsics instead since they are supported directly by the AMDGPU backend.

### CODE REVIEW

File: mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp [hunk 1]
```diff
@@ -84,16 +86,12 @@ void mlir::populateMathToROCDLConversionPatterns(LLVMTypeConverter &converter,
...
                                    "__ocml_sinh_f64");
-  populateOpPatterns<math::ExpOp>(converter, patterns, "__ocml_exp_f32",
-                                  "__ocml_exp_f64");
   populateOpPatterns<math::Exp2Op>(converter, patterns, "__ocml_exp2_f32",
                                    "__ocml_exp2_f64");
   populateOpPatterns<math::ExpM1Op>(converter, patterns, "__ocml_expm1_f32",
                                     "__ocml_expm1_f64");
   populateOpPatterns<math::FloorOp>(converter, patterns, "__ocml_floor_f32",
                                     "__ocml_floor_f64");
-  populateOpPatterns<math::LogOp>(converter, patterns, "__ocml_log_f32",
```
Comment (krzysz00) [inline mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp:95]: Same note re f64 log

File: mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp [hunk 2]
```diff
@@ -84,16 +86,12 @@ void mlir::populateMathToROCDLConversionPatterns(LLVMTypeConverter &converter,
                                    "__ocml_cosh_f64");
   populateOpPatterns<math::SinhOp>(converter, patterns, "__ocml_sinh_f32",
                                    "__ocml_sinh_f64");
-  populateOpPatterns<math::ExpOp>(converter, patterns, "__ocml_exp_f32",
```
Comment (krzysz00) [inline mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp:87]: Having disassembled OCML, the double-precision exp isn't actually a direct wrapper around the relevant intrinsic, but I figure that's probably fine

**Review by krzysz00 — COMMENTED** [review]
I *think* this is fine, but I want to double-check that nothing'll go wrong with double-precision `exp` and `log`

File: mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp (hunk 2 above)
Comment (arsenm) [reply mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp:87]: Correct. We only directly handle the f32 (and I think f16) versions. The f64 versions of the hard operations do not work. We do directly handle llvm.sqrt.f64 as an exception

Comment (jsjodin) [reply mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp:87]: So what do we do in the case of f32 being handled but f64 not, should we still just call ocml for both or modify the lowering to handle just one of them?

Comment (arsenm) [reply mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp:87]: Modify the lowering to handle just one. The operation + type should be treated like different operations, so emit the working f32 intrinsics and the calls for the nonworking f64

Comment (jsjodin) [reply mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp:87]: Okay, I put the lowering for f64 back for log and exp and added the tests back for them as well.

**Review by krzysz00 — APPROVED** [review]
Approved

(If it's possible, could you get the half-precision functions hooked up in a later PR?)

**Comment by jsjodin** [comment]
> Approved
> 
> (If it's possible, could you get the half-precision functions hooked up in a later PR?)

Sure, that should be pretty easy to do.

**Comment by akuegel** [comment]
This patch now breaks lowering for bf16 ExpOp or LogOp. For bf16, we upcast to F32, and now we have no lowering for F32 anymore.
Moreover I wonder whether it actually makes a difference in the final lowering? If there are intrinsics, shouldn't the other pattern also eventually lower to them?

**Comment by akuegel** [comment]
@krzysz00 XLA is triggering this bug that bf16 ExpOp and LogOp cannot be lowered anymore: https://github.com/openxla/xla/issues/19700

**Comment by krzysz00** [comment]
@akuegel You'll want to run `--arith-emulate-unsupported-floats` to expand out the bf16 operations, I think.

Alternatively, I'll take a followup patch to lower BF16 `math.*` to the equivalent f32 calls (LLVM intrinsics or OCML ones as relevant) with `extf` and `truncf`

**Comment by krzysz00** [comment]
That is to say, there is no bf16 `exp` or `log` and they need to be rewritten to their f32 counterparts

**Comment by akuegel** [comment]
Let me expand a bit. OpToFuncCallLowering (which is the shared logic with the NVVM conversion) upcasts bf16 to f32 before checking whether there is a device function available:

https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/GPUCommon/OpToFuncCallLowering.h#L71

Previously, this worked fine, as when upcasting ExpOp (or LogOp) from bf16 to f32, there was a corresponding f32 function specified. Now, this doesn't work anymore, as there is no corresponding f32 function specified. Users of this pass now need to upcast bf16 ops earlier. Is this really what we want?

Let me repeat my earlier question: Does this patch actually make a difference in the final lowering of math::ExpOp for f32? If there are intrinsics, shouldn't the other pattern also eventually lower to them? And if not, wouldn't the right fix be to change the ocml functions to make use of the intrinsics? @krzysz00 @jsjodin @arsenm maybe one of you can answer this?

**Comment by krzysz00** [comment]
This patch does change the lowering to use the compiler intrinsics. The OCML functions in question were, for f32, redundant wrappers for the intrinsics that the compiler team wants to remove.

For bf16 exp and log, I'd almost think that an LLVM patch is in order

**Comment by akuegel** [comment]
For XLA, there is now https://github.com/openxla/xla/pull/19913 which should fix the issue by upcasting log and exp ops with bf16 type early.
I will leave it to people from AMD to figure out whether that is the permanent solution or not, I just wanted to point out the (possibly unintended) side effect of this patch.

**Comment by krzysz00** [comment]
@arsenm and other folks on the compiler side - would there be any reason not to expand exp and log on bfloats during SelectionDAG/GISel?

**Comment by arsenm** [comment]
> @arsenm and other folks on the compiler side - would there be any reason not to expand exp and log on bfloats during SelectionDAG/GISel?

No. In principle all the math intrinsics should be legalized for all types

**Comment by krzysz00** [comment]
Ok, so maybe I'm misreading the above comments and we need an MLIR-side patch to make sure that the bf16 `math.exp` and `math.log` lower to their LLVM intrinsics

**Pull request merged**